apps/frontend/core-dapp/src/stores/entities/FeeEntity.ts

```diff
@@ -0,0 +1,118 @@
+import { BigNumber } from 'ethers'
+import memoize from 'lodash/memoize'
+import { makeAutoObservable } from 'mobx'
+import { formatEther } from 'ethers/lib/utils'
+import { TokenSenderEntity } from './TokenSenderEntity'
+import { Fraction, mul, toPercent } from '../../utils/fraction-utils'
+
+export type PpoReward = {
+  readonly inEth: BigNumber
+  readonly inPpo: BigNumber
+}
+
+export const EMPTY_PPO_REBATE: PpoReward = {
+  inEth: BigNumber.from(0),
+  inPpo: BigNumber.from(0),
+}
+
+type DebugInfo = {
+  tokenSender: string
+  fee: string
+  ppo: string
+  eth: string
+}
+
+const printDebugInfo = memoize((info: DebugInfo[]): void => {
+  if (info.length <= 1) return
+  // eslint-disable-next-line no-console
+  console.table(info)
+}, JSON.stringify)
+
+export class FeeStore {
+  private readonly applicableToWstEthGetter: () => BigNumber | undefined
+  private readonly tokenSender: TokenSenderEntity
+
+  readonly fee: Fraction
+
+  constructor({
+    applicableToWstEth,
+    fee,
+    tokenSender,
+  }: {
+    applicableToWstEth: () => BigNumber | undefined
+    fee: Fraction
+    tokenSender: TokenSenderEntity
+  }) {
+    this.applicableToWstEthGetter = applicableToWstEth
+    this.fee = fee
+    this.tokenSender = tokenSender
+
+    makeAutoObservable(this)
+  }
+
+  get amountAfterFeeInWstEth(): BigNumber | undefined {
+    const { feeAmountInWstEth } = this
+    const applicableTo = this.applicableToWstEthGetter()
+    if (applicableTo === undefined || feeAmountInWstEth === undefined) return undefined
+    return applicableTo.sub(feeAmountInWstEth)
+  }
+
+  get feeAmountInWstEth(): BigNumber | undefined {
+    const applicableTo = this.applicableToWstEthGetter()
+    if (applicableTo === undefined) return undefined
+    return mul(applicableTo, this.fee)
+  }
+
+  static getTotalFeeInWstEth(fees: readonly FeeStore[]): BigNumber | undefined {
+    let total = BigNumber.from(0)
+
+    for (let i = 0; i < fees.length; i++) {
+      const fee = fees[i]
+      if (fee.feeAmountInWstEth === undefined) return undefined
+      total = total.add(fee.feeAmountInWstEth)
+    }
```

Comment:
would prefer that we use `forEach` if possible, for better code readability and also less room for mistake
```suggestion
    fees.forEach((fee) => {
      if (fee.feeAmountInWstEth === undefined) return undefined
      total = total.add(fee.feeAmountInWstEth)
    })
```

Reply:
That wouldn't work because the `return undefined` is intended to return from the entire `getTotalFeeInWstEth`, and here it would return from the inline `forEach` function instead.

I tried to do `for..of` for better readability but our [`no-restricted-syntax` ESLint rule](https://eslint.org/docs/latest/rules/no-restricted-syntax) prevents that.

Reply:
oh you're right, it needs to return, np!